### PR TITLE
stop and cleanup tracer on constructor error

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Restructures the tracer constructor to cleanup dependencies in case of any error during the function.

### Motivation

One test failure due to some error during this constructor resulted in panics on later tests, due to duplicate collector registration.

### Additional Notes

The events registration was changed from a function to an interface, because functions are not comparable and thus could not be unregistered.

The `go.mod 1.20` change was required for interfaces to be comparable. Once #18874 is merged, I will remove it.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by akarpz — APPROVED** [review]

File: pkg/network/events/monitor.go [hunk 1]
```diff
@@ -35,15 +36,22 @@ func Init() error {
 	return initErr
 }
 
-// HandlerFunc is the prototype for an event handler callback for process events
-type HandlerFunc func(*model.ProcessContext)
+type ProcessEventHandler interface {
+	HandleProcessEvent(*model.ProcessContext)
```
Comment (hmahmood) [inline pkg/network/events/monitor.go:40]: Just `Handle`?

File: pkg/network/tracer/tracer.go [hunk 2]
```diff
@@ -119,7 +117,7 @@ func NewTracer(config *config.Config) (*Tracer, error) {
 
 // newTracer is an internal function used by tests primarily
 // (and NewTracer above)
-func newTracer(cfg *config.Config) (*Tracer, error) {
+func newTracer(cfg *config.Config) (rettr *Tracer, reterr error) {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer.go:120]: I don't see `rettr` used anywhere, and `reterr` seems to not be set anywhere.

File: pkg/network/tracer/tracer.go [hunk 3]
```diff
@@ -161,48 +153,42 @@ func newTracer(cfg *config.Config) (*Tracer, error) {
...
+
+	tr.ebpfTracer, err = connection.NewTracer(cfg, tr.bpfTelemetry)
 	if err != nil {
 		return nil, err
 	}
-	coretelemetry.GetCompatComponent().RegisterCollector(ebpfTracer)
+	coretelemetry.GetCompatComponent().RegisterCollector(tr.ebpfTracer)
 
-	conntracker, err := newConntracker(cfg, bpfTelemetry)
+	tr.conntracker, err = newConntracker(cfg, tr.bpfTelemetry)
```
Comment (hmahmood) [inline pkg/network/tracer/tracer.go:179]: Can `tr.conntracker` be `nil` here?

File: pkg/network/tracer/tracer.go [hunk 4]
```diff
@@ -353,13 +348,28 @@ func (t *Tracer) Stop() {
...
 		t.gwLookup.Close()
 	}
-	t.reverseDNS.Close()
-	t.ebpfTracer.Stop()
-	t.usmMonitor.Stop()
-	t.conntracker.Close()
-	t.processCache.Stop()
-	close(t.exitTelemetry)
-	coretelemetry.GetCompatComponent().Reset()
+	if t.reverseDNS != nil {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer.go:351]: Do we need to check for `nil` for these? I thought most of these objects have a check for `nil` in their `Close` methods.

File: pkg/network/tracer/tracer.go (hunk 2 above)
Comment (brycekahle) [reply pkg/network/tracer/tracer.go:120]: `reterr` is automatically set if you use `return`

File: pkg/network/events/monitor.go (hunk 1 above)
Comment (brycekahle) [reply pkg/network/events/monitor.go:40]: That seems pretty generic for an interface method. The function can be part of a object which does a lot of things, so `Handle` doesn't indicate much.

File: pkg/network/tracer/tracer.go (hunk 3 above)
Comment (brycekahle) [reply pkg/network/tracer/tracer.go:179]: What do you mean? It is being assigned to.

File: pkg/network/tracer/tracer.go (hunk 4 above)
Comment (brycekahle) [reply pkg/network/tracer/tracer.go:351]: I'm not sure if they do, and that seems like a leaky abstraction. You have to know that the function will handle being called on a `nil` object, which is not idiomatic Go.

File: pkg/network/tracer/tracer.go (hunk 2 above)
Comment (brycekahle) [reply pkg/network/tracer/tracer.go:120]: Also, if you name one return value, you have to name them all.

Comment (hmahmood) [reply pkg/network/tracer/tracer.go:120]: I think `(_ *Tracer, reterr error)` works.

File: pkg/network/tracer/tracer.go (hunk 3 above)
Comment (hmahmood) [reply pkg/network/tracer/tracer.go:179]: You check for `tr.conntracker != nil` in `Stop()`, but assume that is going to be `non-nil` after this line. Can `tr.conntracker` ever be non-`nil` if we don't return an error from `newTracer`?

Comment (brycekahle) [reply pkg/network/tracer/tracer.go:179]: if `newConntracker` returns an error, `tr.conntracker` will be `nil`.

**Review by ogaca-dd — APPROVED** [review]
LGTM for files owned by agent shared components

**Pull request merged**